extensions/protobuf/src/main/java/io/deephaven/protobuf/ProtobufDescriptorParserImpl.java

```diff
@@ -0,0 +1,545 @@
+        }
+
+        private ProtobufFunctions functions() {
+            return functions(false);
+        }
+
+        private ProtobufFunctions functions(boolean forceInclude) {
+            final FieldOptions fo = options.fieldOptions().apply(fieldPath);
+            if (!forceInclude && !fo.include()) {
+                return ProtobufFunctions.empty();
+            }
+            final ProtobufFunctions wellKnown = wellKnown().orElse(null);
+            if (wellKnown != null && fo.wellKnown() == WellKnownBehavior.asWellKnown()) {
+                return wellKnown;
+            }
+            if (fd.isMapField() && fo.map() == MapBehavior.asMap()) {
+                return new MapFieldObject().functions();
+            }
+            if (fd.isRepeated()) {
+                return new RepeatedFieldObject().functions();
+            }
+            return new FieldObject().functions();
+        }
+
+        private Optional<ProtobufFunctions> wellKnown() {
+            // todo: eventually have support for parsing specific fields in specific ways
+            return Optional.empty();
+        }
+
+        private ProtobufFunctions namedField(TypedFunction<Message> tf) {
+            // todo: can we re-work this, so we can use fieldPath here instead and not use prefix later?
+            return ProtobufFunctions.of(ProtobufFunction.of(FieldPath.of(fd), tf));
+        }
+
+        private DescriptorContext toMessageContext() {
+            if (fd.getJavaType() != JavaType.MESSAGE) {
+                throw new IllegalStateException();
+            }
+            return new DescriptorContext(fieldPath, fd.getMessageType());
+        }
+
+        private class FieldObject implements ObjectFunction<Message, Object> {
+
+            @Override
+            public GenericType<Object> returnType() {
+                return Type.ofCustom(Object.class);
+            }
+
+            @Override
+            public Object apply(Message message) {
+                // Note: in protobuf an actualized Message is never null - getField will always return non-null (and
+                // thus, sub Messages will never be natively null).
+                //
+                // In the case of our translation layer though, the presence of a null Message means that this field, or
+                // some parent of this field, was "not present".
+                if (message == null) {
+                    return null;
+                }
+                // hasField only semantically meaningful when hasPresence == true
+                if (fd.hasPresence() && !message.hasField(fd)) {
+                    return null;
+                }
+                return message.getField(fd);
+            }
+
+            private ProtobufFunctions functions() {
+                // Note: we might be tempted at this layer to treat null boxed primitives as DH null primitives, but
+                // 1) this parsing layer doesn't / shouldn't need to know about DH nulls
```

Comment:
So; at the end of the day, the functional composition is going to look exactly the same / perform exactly the same (there won't be any extraneous steps along the functional chain), it's just a matter of "who is responsible"?

In this specific example, we have:

```java
  switch (fd.getJavaType()) {
      case INT:
          return fd.hasPresence()
                  // ToObjectFunction<Message, Integer>
                  ? namedField(mapToObj(BOXED_INT_OBJ))
                  // ToIntFunction<Message>
                  : namedField(mapToInt(ToIntFunction.cast()));
```

if we were handling DH null at this layer, we could do:

```java
    switch (fd.getJavaType()) {
        case INT:
            return fd.hasPresence()
                    // ToIntFunction<Message>
                    ? namedField(mapToObj(BOXED_INT_OBJ).mapToInt(ProtobufDescriptorParserImpl::unbox))
                    // ToIntFunction<Message>
                    : namedField(mapToInt(ToIntFunction.cast())); 
...
    private static int unbox(Integer x) {
        return x == null ? QueryConstants.NULL_INT : x;
    }
```

But my claim is that this sort of transform from `ToObjectFunction<X, Integer>` -> `ToIntFunction<X>` is better left to high layers / callers of this code so that they get the additional context of "this field was explicitly optional", and they can apply the equivalent of `mapToInt(ProtobufDescriptorParserImpl::unbox)` themselves (ie, `UnboxTransform` in the kafka layer).

There _is_ a bit of nuance though, b/c I've agreed w/ your take with respect to nullability in `BypassOnNull`, which makes sure we don't unnecessarily _box_ from int -> Integer just to preserve the nullability context.